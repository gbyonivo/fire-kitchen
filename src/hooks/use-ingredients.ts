import {
  setFetchingIngredients,
  setFetchingIngredientsError,
  setIngredients,
} from "@/lib/slices/cache-slice";
import { RootState } from "@/lib/store";
import {
  Ingredient,
  isSampleIngredient,
  SampleIngredient,
} from "@/types/ingredient";
import { convertIngredientList } from "@/utils/ingridient-converter";
import { KitchenAxios } from "@/utils/kitchen-axios";
import { AxiosError } from "axios";
import { useCallback, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";

export function useIngredients(): {
  ingredients: Ingredient[];
  fetching: boolean;
  error: AxiosError | string | null;
  fetchIngredients: () => void;
  ingridientCache: Record<string, Ingredient>;
} {
  const {
    ingredients,
    fetchingIngredients,
    fetchingIngredientsError,
    ingridientCache,
  } = useSelector((state: RootState) => state.cache);
  const dispatch = useDispatch();

  const fetchIngredients = useCallback(() => {
    dispatch(setFetchingIngredients());
    const fetch = async () => {
      try {
        const response = await KitchenAxios.get("list.php?i=list");
        const data = (await response.data) as { meals: unknown[] };

        if (isSampleIngredient(data.meals?.[0])) {
          dispatch(
            setIngredients(
              convertIngredientList(data.meals as unknown as SampleIngredient[])
            )
          );
        } else {
          dispatch(setFetchingIngredientsError("Invalid data"));
        }
      } catch (e) {
        dispatch(setFetchingIngredientsError(e as AxiosError));
      }
    };
    fetch();
  }, [dispatch]);

  useEffect(() => {
    if (ingredients.length > 0 || fetchingIngredients) return;
    fetchIngredients();
  }, [fetchIngredients, ingredients, fetchingIngredients]);

  return {
    ingredients,
    fetchIngredients,
    fetching: fetchingIngredients,
    error: fetchingIngredientsError,
    ingridientCache,
  };
}
