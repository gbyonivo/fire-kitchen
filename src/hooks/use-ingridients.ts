import {
  setFetchingIngridients,
  setFetchingIngridientsError,
  setIngridients,
} from "@/lib/slices/cache-slice";
import { RootState } from "@/lib/store";
import {
  Ingredient,
  isSampleIngredient,
  SampleIngredient,
} from "@/types/ingredient";
import { convertIngredientList } from "@/utils/ingridient-converter";
import { KitchenAxios } from "@/utils/kitchen-axios";
import { AxiosError } from "axios";
import { useCallback, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";

export function useIngridients(): {
  ingridients: Ingredient[];
  fetching: boolean;
  error: AxiosError | string | null;
  fetchIngridients: () => void;
  ingridientCache: Record<string, Ingredient>;
} {
  const {
    ingridients,
    fetchingIngridients,
    fetchingIngridientsError,
    ingridientCache,
  } = useSelector((state: RootState) => state.cache);
  const dispatch = useDispatch();

  const fetchIngridients = useCallback(() => {
    console.log("fetching ingridients 1");
    dispatch(setFetchingIngridients());
    const fetch = async () => {
      console.log("fetching ingridients 2");
      try {
        const response = await KitchenAxios.get("list.php?i=list");
        const data = (await response.data) as { meals: unknown[] };

        if (isSampleIngredient(data.meals?.[0])) {
          dispatch(
            setIngridients(
              convertIngredientList(data.meals as unknown as SampleIngredient[])
            )
          );
        } else {
          dispatch(setFetchingIngridientsError("Invalid data"));
        }
      } catch (e) {
        dispatch(setFetchingIngridientsError(e as AxiosError));
      }
    };
    fetch();
  }, [dispatch]);

  useEffect(() => {
    if (ingridients.length > 0 || fetchingIngridients) return;
    console.log("fetching ingridients");
    fetchIngridients();
  }, [fetchIngridients, ingridients, fetchingIngridients]);

  return {
    ingridients,
    fetchIngridients,
    fetching: fetchingIngridients,
    error: fetchingIngridientsError,
    ingridientCache,
  };
}
