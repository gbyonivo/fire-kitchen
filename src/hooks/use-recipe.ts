import { RootState } from "@/lib/store";
import { useCallback, useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { AxiosError } from "axios";
import { KitchenAxios } from "@/utils/kitchen-axios";
import { convertRecipe } from "@/utils/recipe-converter";
import { updateRecipeCache } from "@/lib/slices/cache-slice";

export function useRecipe({ recipeId }: { recipeId: string }) {
  const dispatch = useDispatch();
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<AxiosError | null>(null);

  const recipeCache = useSelector(
    (state: RootState) => state.cache.recipeCache
  );
  const recipe = recipeCache?.[recipeId];

  const fetchRecipe = useCallback(() => {
    const getRecipe = async () => {
      if (recipe) return;
      try {
        setLoading(true);
        const response = await KitchenAxios.get(`/lookup.php`, {
          params: {
            i: recipeId,
          },
        });
        const recipe = convertRecipe(response.data.meals[0]);
        if (recipe) {
          dispatch(updateRecipeCache(recipe));
        } else {
          setError(new AxiosError("Recipe not found"));
        }
        setLoading(false);
      } catch (error) {
        setLoading(false);
        setError(error as AxiosError);
      }
    };
    getRecipe();
  }, [dispatch, recipeId, recipe]);

  useEffect(() => {
    if (recipe?.id || loading || error) return;
    fetchRecipe();
  }, [fetchRecipe, recipe?.id, loading, error]);

  return {
    recipe,
    fetchRecipe,
    loading,
    error,
  };
}
