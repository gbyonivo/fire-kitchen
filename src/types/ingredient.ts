export interface SampleIngredient {
  idIngredient?: string;
  strIngredient?: string;
  strDescription?: string;
  strType?: string;
}

export function isSampleIngredient(
  ingredient: unknown
): ingredient is SampleIngredient {
  return (
    typeof ingredient === "object" &&
    ingredient !== null &&
    "idIngredient" in ingredient &&
    "strIngredient" in ingredient
  );
}

export interface Ingredient {
  id: string;
  ingredient: string;
  description?: string;
  type?: string;
}
