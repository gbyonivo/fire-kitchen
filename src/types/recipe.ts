export interface RecipeIngredient {
  name: string;
  measure: string;
}

export interface Recipe {
  id: string;
  name: string;
  category: string;
  area: string;
  instructions: string;
  thumbnail: string;
  tags: string[];
  youtube: string;
  ingredients: RecipeIngredient[];
}
