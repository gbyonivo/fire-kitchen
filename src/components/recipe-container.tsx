"use client";

import { useRecipe } from "@/hooks/use-recipe";
import { useParams } from "next/navigation";
import { Spinner } from "./common/spinner";
import Image from "next/image";
import { ArrowTopRightOnSquareIcon } from "@heroicons/react/24/solid";

export function RecipeContainer() {
  const { id } = useParams();
  const { recipe, loading, error } = useRecipe({ recipeId: id as string });

  console.log(recipe, loading, error);
  return (
    <div className="flex w-screen h-screen justify-center py-16">
      {loading && <Spinner size="large" />}
      {error && <div>Error: {error.message}</div>}
      {recipe && (
        <div className="flex flex-col md:w-2/3 md:overflow-hidden">
          <h1 className="text-2xl font-bold text-center mb-8">{recipe.name}</h1>
          <div className="flex w-full flex-col md:flex-row">
            <div className="w-full md:w-2/6">
              <div className="flex justify-center">
                <Image
                  src={recipe.thumbnail}
                  alt={recipe.name}
                  width={200}
                  height={200}
                  className="self-center rounded-lg"
                />
              </div>
              <div className="flex flex-col my-4 gap-4">
                <p className="text-sm text-center">{recipe.category}</p>
                <p className="text-sm text-center">{recipe.area}</p>
                <p className="text-sm text-center">
                  <a
                    href={recipe.youtube}
                    target="_blank"
                    rel="noreferrer"
                    className="underline"
                  >
                    Watch on Youtube{" "}
                    <ArrowTopRightOnSquareIcon className="w-4 h-4 inline ml-2" />
                  </a>
                </p>
              </div>
            </div>
            <div className="w-full md:w-4/6 md:h-5/6 border-l border-gray-300 px-8 overflow-y-auto">
              <div className="flex flex-col gap-4">
                <p className="text-sm font-bold text-gray-500">Ingredients</p>
                <div className="grid grid-cols-2 gap-2">
                  {recipe.ingredients.map((ingredient) => (
                    <span key={ingredient.name} className="text-sm text-left">
                      {ingredient.measure} {ingredient.name}
                    </span>
                  ))}
                </div>
                <p className="text-sm font-bold text-gray-500">Instructions</p>
                <p className="text-sm text-left pb-24">{recipe.instructions}</p>
              </div>
            </div>
          </div>
        </div>
      )}
    </div>
  );
}
