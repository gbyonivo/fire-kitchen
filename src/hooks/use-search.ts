import { searchSuccess, setSearching } from "@/lib/slices/search-slice";
import { Ingredient } from "@/types/ingredient";
import { KitchenAxios } from "@/utils/kitchen-axios";
import { convertRecipe } from "@/utils/recipe-converter";
import { useCallback } from "react";
import { useDispatch } from "react-redux";

export function useSearch() {
  const dispatch = useDispatch();

  const searchRecipe = useCallback(
    async (ingredients: Ingredient[], search: string) => {
      try {
        dispatch(setSearching());
        const searches = [];
        if (ingredients.length > 0) {
          searches.push(
            KitchenAxios.get("filter.php", {
              params: {
                i: ingredients
                  .map((ingredient) => ingredient.ingredient)
                  .join(","),
              },
            })
          );
        }
        if (search.length > 2) {
          searches.push(
            KitchenAxios.get("/search.php", {
              params: {
                s: search,
              },
            })
          );
        }

        const responses = await Promise.all(searches);
        const recipes = responses.flatMap(
          (response) => response.data.meals || []
        );
        const uniqueRecipes = recipes.reduce((acc, recipe) => {
          if (!recipe?.idMeal) {
            return acc;
          }
          return {
            ...acc,
            [recipe.idMeal]: convertRecipe(recipe),
          };
        }, {});

        console.log(recipes.length, Object.values(uniqueRecipes));

        dispatch(searchSuccess(Object.values(uniqueRecipes)));
      } catch (error) {
        console.error(error);
      }
    },
    [dispatch]
  );

  return { searchRecipe };
}
