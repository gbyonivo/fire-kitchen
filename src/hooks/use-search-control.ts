import { useState } from "react";

import { useRouter, useSearchParams } from "next/navigation";

import { usePathname } from "next/navigation";
import { useSearch } from "./use-search";
import { createQueryString } from "@/utils/common";
import { useEffect } from "react";
import { Ingredient } from "@/types/ingredient";
import { useIngredients } from "./use-ingredients";

export const useSearchControl = () => {
  const pathname = usePathname();
  const router = useRouter();
  const searchParams = useSearchParams();
  const { ingridientCache } = useIngredients();
  const { searchRecipe } = useSearch();
  const getIngredients = () => {
    return (
      searchParams
        .get("ingredients")
        ?.split(",")
        .map((ingredient) => ingridientCache?.[ingredient]) || []
    );
  };

  const [usedParams, setUsedParams] = useState<{
    search: string;
    ingredients: Ingredient[];
  }>(() => ({
    search: searchParams.get("search") || "",
    ingredients: getIngredients(),
  }));

  useEffect(() => {
    router.push(
      pathname +
        "?" +
        createQueryString({
          nameValuePairs: [
            { name: "search", value: usedParams.search },
            {
              name: "ingredients",
              value: usedParams.ingredients.map((i) => i.ingredient).join(","),
            },
          ],
          searchParams,
        })
    );
    searchRecipe(usedParams.ingredients, usedParams.search);
  }, [pathname, router, searchParams, searchRecipe, usedParams]);

  return {
    onSetParams: (params: { search: string; ingredients: Ingredient[] }) => {
      setUsedParams(params);
    },
    initialSearchValue: searchParams.get("search") || "",
    initialIngredients: getIngredients(),
  };
};
