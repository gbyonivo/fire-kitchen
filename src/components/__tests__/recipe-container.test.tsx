import { render, screen, waitFor } from "@testing-library/react";
import { Recipe } from "@/types/recipe";
import { useRecipe } from "@/hooks/use-recipe";
import { configureStore } from "@reduxjs/toolkit";
import { RecipeContainer } from "../recipe-container";
import { Provider } from "react-redux";
import cacheReducer from "@/lib/slices/cache-slice";
import searchReducer from "@/lib/slices/search-slice";
import { useParams } from "next/navigation";
import userEvent from "@testing-library/user-event";

const store = configureStore({
  reducer: {
    cache: cacheReducer,
    search: searchReducer,
  },
});

jest.mock("@/hooks/use-recipe");
jest.mock("next/navigation", () => ({
  useParams: jest.fn().mockReturnValue({ id: "1" }),
}));

const useRecipeReturnValue: {
  recipe: Recipe;
  loading: boolean;
  error: Error | null;
} = {
  recipe: {
    id: "1",
    name: "Test",
    ingredients: [
      {
        name: "Name Test",
        measure: "Measure Test",
      },
    ],
    category: "Test",
    area: "Test",
    instructions: "Test",
    thumbnail: "/images/test.jpg",
    tags: [],
    youtube: "Test",
  },
  loading: false,
  error: null,
};

describe("RecipeContainer", () => {
  test.each([
    [useRecipeReturnValue],
    [
      {
        ...useRecipeReturnValue,
        loading: true,
        error: null,
      },
    ],
    [
      {
        ...useRecipeReturnValue,
        loading: false,
        error: new Error("Test"),
      },
    ],
    [
      {
        ...useRecipeReturnValue,
        recipe: null,
        loading: false,
        error: new Error("Test"),
      },
    ],
    [
      {
        ...useRecipeReturnValue,
        recipe: null,
        loading: true,
        error: false,
      },
    ],
  ])("%s should match snapshot", async (returnValue) => {
    (useRecipe as unknown as jest.Mock).mockReturnValue(returnValue);
    (useParams as unknown as jest.Mock).mockReturnValue({ id: "1" });
    const { container } = render(
      <Provider store={store}>
        <RecipeContainer />
      </Provider>
    );
    await userEvent.hover(screen.getByTestId("recipe-container"));
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});
