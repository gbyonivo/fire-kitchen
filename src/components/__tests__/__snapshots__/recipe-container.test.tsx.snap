// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeContainer {
  recipe: [Object],
  loading: false,
  error: Error: Test
      at /Users/orkumaivo/workspace/fire-kitchen/src/components/__tests__/recipe-container.test.tsx:63:16
      at _dispatchDescribe (/Users/orkumaivo/workspace/fire-kitchen/node_modules/jest-circus/build/index.js:91:26)
      at describe (/Users/orkumaivo/workspace/fire-kitchen/node_modules/jest-circus/build/index.js:55:5)
      at Object.<anonymous> (/Users/orkumaivo/workspace/fire-kitchen/src/components/__tests__/recipe-container.test.tsx:49:1)
      at Runtime._execModule (/Users/orkumaivo/workspace/fire-kitchen/node_modules/jest-runtime/build/index.js:1439:24)
      at Runtime._loadModule (/Users/orkumaivo/workspace/fire-kitchen/node_modules/jest-runtime/build/index.js:1022:12)
      at Runtime.requireModule (/Users/orkumaivo/workspace/fire-kitchen/node_modules/jest-runtime/build/index.js:882:12)
      at jestAdapter (/Users/orkumaivo/workspace/fire-kitchen/node_modules/jest-circus/build/legacy-code-todo-rewrite/jestAdapter.js:77:13)
      at async runTestInternal (/Users/orkumaivo/workspace/fire-kitchen/node_modules/jest-runner/build/runTest.js:367:16)
      at async runTest (/Users/orkumaivo/workspace/fire-kitchen/node_modules/jest-runner/build/runTest.js:444:34)
} should match snapshot 1`] = `
<div>
  <div
    class="flex w-screen h-screen justify-center py-16"
    data-testid="recipe-container"
  >
    <div>
      Error: 
      Test
    </div>
    <div
      class="flex flex-col md:w-2/3 md:overflow-hidden"
    >
      <h1
        class="text-2xl font-bold text-center mb-8"
      >
        Test
      </h1>
      <div
        class="flex w-full flex-col md:flex-row"
      >
        <div
          class="w-full md:w-2/6"
        >
          <div
            class="flex justify-center"
          >
            <img
              alt="Test"
              class="self-center rounded-lg"
              data-nimg="1"
              decoding="async"
              height="200"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2Ftest.jpg&w=640&q=75"
              srcset="/_next/image?url=%2Fimages%2Ftest.jpg&w=256&q=75 1x, /_next/image?url=%2Fimages%2Ftest.jpg&w=640&q=75 2x"
              style="color: transparent;"
              width="200"
            />
          </div>
          <div
            class="flex flex-col my-4 gap-4"
          >
            <p
              class="text-sm text-center"
            >
              Test
            </p>
            <p
              class="text-sm text-center"
            >
              Test
            </p>
            <p
              class="text-sm text-center"
            >
              <a
                class="underline"
                href="Test"
                rel="noreferrer"
                target="_blank"
              >
                Watch on Youtube
                 
                <svg
                  aria-hidden="true"
                  class="w-4 h-4 inline ml-2"
                  data-slot="icon"
                  fill="currentColor"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M15.75 2.25H21a.75.75 0 0 1 .75.75v5.25a.75.75 0 0 1-1.5 0V4.81L8.03 17.03a.75.75 0 0 1-1.06-1.06L19.19 3.75h-3.44a.75.75 0 0 1 0-1.5Zm-10.5 4.5a1.5 1.5 0 0 0-1.5 1.5v10.5a1.5 1.5 0 0 0 1.5 1.5h10.5a1.5 1.5 0 0 0 1.5-1.5V10.5a.75.75 0 0 1 1.5 0v8.25a3 3 0 0 1-3 3H5.25a3 3 0 0 1-3-3V8.25a3 3 0 0 1 3-3h8.25a.75.75 0 0 1 0 1.5H5.25Z"
                    fill-rule="evenodd"
                  />
                </svg>
              </a>
            </p>
          </div>
        </div>
        <div
          class="w-full md:w-4/6 md:h-5/6 border-l border-gray-300 px-8 overflow-y-auto"
        >
          <div
            class="flex flex-col gap-4 mb-8"
          >
            <p
              class="text-sm font-bold text-gray-500"
            >
              Ingredients
            </p>
            <div
              class="grid grid-cols-2 gap-2"
            >
              <span
                class="text-sm text-left text-gray-100/50"
                data-testid="ingredient-Name Test-Measure Test"
                role="button"
              >
                Measure Test
                 
                Name Test
              </span>
            </div>
          </div>
          <p
            class="text-sm font-bold text-gray-500"
          >
            Instructions
          </p>
          <p
            class="text-sm text-left pb-24"
          >
            Test
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`RecipeContainer {
  recipe: null,
  loading: false,
  error: Error: Test
      at /Users/orkumaivo/workspace/fire-kitchen/src/components/__tests__/recipe-container.test.tsx:71:16
      at _dispatchDescribe (/Users/orkumaivo/workspace/fire-kitchen/node_modules/jest-circus/build/index.js:91:26)
      at describe (/Users/orkumaivo/workspace/fire-kitchen/node_modules/jest-circus/build/index.js:55:5)
      at Object.<anonymous> (/Users/orkumaivo/workspace/fire-kitchen/src/components/__tests__/recipe-container.test.tsx:49:1)
      at Runtime._execModule (/Users/orkumaivo/workspace/fire-kitchen/node_modules/jest-runtime/build/index.js:1439:24)
      at Runtime._loadModule (/Users/orkumaivo/workspace/fire-kitchen/node_modules/jest-runtime/build/index.js:1022:12)
      at Runtime.requireModule (/Users/orkumaivo/workspace/fire-kitchen/node_modules/jest-runtime/build/index.js:882:12)
      at jestAdapter (/Users/orkumaivo/workspace/fire-kitchen/node_modules/jest-circus/build/legacy-code-todo-rewrite/jestAdapter.js:77:13)
      at async runTestInternal (/Users/orkumaivo/workspace/fire-kitchen/node_modules/jest-runner/build/runTest.js:367:16)
      at async runTest (/Users/orkumaivo/workspace/fire-kitchen/node_modules/jest-runner/build/runTest.js:444:34)
} should match snapshot 1`] = `
<div>
  <div
    class="flex w-screen h-screen justify-center py-16"
    data-testid="recipe-container"
  >
    <div>
      Error: 
      Test
    </div>
  </div>
</div>
`;

exports[`RecipeContainer { recipe: [Object], loading: false, error: null } should match snapshot 1`] = `
<div>
  <div
    class="flex w-screen h-screen justify-center py-16"
    data-testid="recipe-container"
  >
    <div
      class="flex flex-col md:w-2/3 md:overflow-hidden"
    >
      <h1
        class="text-2xl font-bold text-center mb-8"
      >
        Test
      </h1>
      <div
        class="flex w-full flex-col md:flex-row"
      >
        <div
          class="w-full md:w-2/6"
        >
          <div
            class="flex justify-center"
          >
            <img
              alt="Test"
              class="self-center rounded-lg"
              data-nimg="1"
              decoding="async"
              height="200"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2Ftest.jpg&w=640&q=75"
              srcset="/_next/image?url=%2Fimages%2Ftest.jpg&w=256&q=75 1x, /_next/image?url=%2Fimages%2Ftest.jpg&w=640&q=75 2x"
              style="color: transparent;"
              width="200"
            />
          </div>
          <div
            class="flex flex-col my-4 gap-4"
          >
            <p
              class="text-sm text-center"
            >
              Test
            </p>
            <p
              class="text-sm text-center"
            >
              Test
            </p>
            <p
              class="text-sm text-center"
            >
              <a
                class="underline"
                href="Test"
                rel="noreferrer"
                target="_blank"
              >
                Watch on Youtube
                 
                <svg
                  aria-hidden="true"
                  class="w-4 h-4 inline ml-2"
                  data-slot="icon"
                  fill="currentColor"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M15.75 2.25H21a.75.75 0 0 1 .75.75v5.25a.75.75 0 0 1-1.5 0V4.81L8.03 17.03a.75.75 0 0 1-1.06-1.06L19.19 3.75h-3.44a.75.75 0 0 1 0-1.5Zm-10.5 4.5a1.5 1.5 0 0 0-1.5 1.5v10.5a1.5 1.5 0 0 0 1.5 1.5h10.5a1.5 1.5 0 0 0 1.5-1.5V10.5a.75.75 0 0 1 1.5 0v8.25a3 3 0 0 1-3 3H5.25a3 3 0 0 1-3-3V8.25a3 3 0 0 1 3-3h8.25a.75.75 0 0 1 0 1.5H5.25Z"
                    fill-rule="evenodd"
                  />
                </svg>
              </a>
            </p>
          </div>
        </div>
        <div
          class="w-full md:w-4/6 md:h-5/6 border-l border-gray-300 px-8 overflow-y-auto"
        >
          <div
            class="flex flex-col gap-4 mb-8"
          >
            <p
              class="text-sm font-bold text-gray-500"
            >
              Ingredients
            </p>
            <div
              class="grid grid-cols-2 gap-2"
            >
              <span
                class="text-sm text-left text-gray-100/50"
                data-testid="ingredient-Name Test-Measure Test"
                role="button"
              >
                Measure Test
                 
                Name Test
              </span>
            </div>
          </div>
          <p
            class="text-sm font-bold text-gray-500"
          >
            Instructions
          </p>
          <p
            class="text-sm text-left pb-24"
          >
            Test
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`RecipeContainer { recipe: [Object], loading: true, error: null } should match snapshot 1`] = `
<div>
  <div
    class="flex w-screen h-screen justify-center py-16"
    data-testid="recipe-container"
  >
    <div>
      <svg
        aria-hidden="true"
        class="animate-spin text-red-500 undefined w-16 h-16"
        data-slot="icon"
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M12.963 2.286a.75.75 0 0 0-1.071-.136 9.742 9.742 0 0 0-3.539 6.176 7.547 7.547 0 0 1-1.705-1.715.75.75 0 0 0-1.152-.082A9 9 0 1 0 15.68 4.534a7.46 7.46 0 0 1-2.717-2.248ZM15.75 14.25a3.75 3.75 0 1 1-7.313-1.172c.628.465 1.35.81 2.133 1a5.99 5.99 0 0 1 1.925-3.546 3.75 3.75 0 0 1 3.255 3.718Z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
    <div
      class="flex flex-col md:w-2/3 md:overflow-hidden"
    >
      <h1
        class="text-2xl font-bold text-center mb-8"
      >
        Test
      </h1>
      <div
        class="flex w-full flex-col md:flex-row"
      >
        <div
          class="w-full md:w-2/6"
        >
          <div
            class="flex justify-center"
          >
            <img
              alt="Test"
              class="self-center rounded-lg"
              data-nimg="1"
              decoding="async"
              height="200"
              loading="lazy"
              src="/_next/image?url=%2Fimages%2Ftest.jpg&w=640&q=75"
              srcset="/_next/image?url=%2Fimages%2Ftest.jpg&w=256&q=75 1x, /_next/image?url=%2Fimages%2Ftest.jpg&w=640&q=75 2x"
              style="color: transparent;"
              width="200"
            />
          </div>
          <div
            class="flex flex-col my-4 gap-4"
          >
            <p
              class="text-sm text-center"
            >
              Test
            </p>
            <p
              class="text-sm text-center"
            >
              Test
            </p>
            <p
              class="text-sm text-center"
            >
              <a
                class="underline"
                href="Test"
                rel="noreferrer"
                target="_blank"
              >
                Watch on Youtube
                 
                <svg
                  aria-hidden="true"
                  class="w-4 h-4 inline ml-2"
                  data-slot="icon"
                  fill="currentColor"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M15.75 2.25H21a.75.75 0 0 1 .75.75v5.25a.75.75 0 0 1-1.5 0V4.81L8.03 17.03a.75.75 0 0 1-1.06-1.06L19.19 3.75h-3.44a.75.75 0 0 1 0-1.5Zm-10.5 4.5a1.5 1.5 0 0 0-1.5 1.5v10.5a1.5 1.5 0 0 0 1.5 1.5h10.5a1.5 1.5 0 0 0 1.5-1.5V10.5a.75.75 0 0 1 1.5 0v8.25a3 3 0 0 1-3 3H5.25a3 3 0 0 1-3-3V8.25a3 3 0 0 1 3-3h8.25a.75.75 0 0 1 0 1.5H5.25Z"
                    fill-rule="evenodd"
                  />
                </svg>
              </a>
            </p>
          </div>
        </div>
        <div
          class="w-full md:w-4/6 md:h-5/6 border-l border-gray-300 px-8 overflow-y-auto"
        >
          <div
            class="flex flex-col gap-4 mb-8"
          >
            <p
              class="text-sm font-bold text-gray-500"
            >
              Ingredients
            </p>
            <div
              class="grid grid-cols-2 gap-2"
            >
              <span
                class="text-sm text-left text-gray-100/50"
                data-testid="ingredient-Name Test-Measure Test"
                role="button"
              >
                Measure Test
                 
                Name Test
              </span>
            </div>
          </div>
          <p
            class="text-sm font-bold text-gray-500"
          >
            Instructions
          </p>
          <p
            class="text-sm text-left pb-24"
          >
            Test
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`RecipeContainer { recipe: null, loading: true, error: false } should match snapshot 1`] = `
<div>
  <div
    class="flex w-screen h-screen justify-center py-16"
    data-testid="recipe-container"
  >
    <div>
      <svg
        aria-hidden="true"
        class="animate-spin text-red-500 undefined w-16 h-16"
        data-slot="icon"
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M12.963 2.286a.75.75 0 0 0-1.071-.136 9.742 9.742 0 0 0-3.539 6.176 7.547 7.547 0 0 1-1.705-1.715.75.75 0 0 0-1.152-.082A9 9 0 1 0 15.68 4.534a7.46 7.46 0 0 1-2.717-2.248ZM15.75 14.25a3.75 3.75 0 1 1-7.313-1.172c.628.465 1.35.81 2.133 1a5.99 5.99 0 0 1 1.925-3.546 3.75 3.75 0 0 1 3.255 3.718Z"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;
